src/OAuth2/Encryption/Jwt.php

```diff
@@ -90,24 +100,17 @@ private function verifySignature($signature, $input, $key, $algo = 'HS256')
 
     private function sign($input, $key, $algo = 'HS256')
     {
-        switch ($algo) {
-            case 'HS256':
-                return hash_hmac('sha256', $input, $key, true);
-
-            case 'HS384':
-                return hash_hmac('sha384', $input, $key, true);
-
-            case 'HS512':
-                return hash_hmac('sha512', $input, $key, true);
+        list($function, $algorithm) = self::$supportedAlgorithms[$algo];
 
-            case 'RS256':
-                return $this->generateRSASignature($input, $key, defined('OPENSSL_ALGO_SHA256') ? OPENSSL_ALGO_SHA256 : 'sha256');
+        switch ($function) {
+            case 'hash_hmac':
+                return hash_hmac($algorithm, $input, $key, true);
 
-            case 'RS384':
-                return $this->generateRSASignature($input, $key, defined('OPENSSL_ALGO_SHA384') ? OPENSSL_ALGO_SHA384 : 'sha384');
-
-            case 'RS512':
-                return $this->generateRSASignature($input, $key, defined('OPENSSL_ALGO_SHA512') ? OPENSSL_ALGO_SHA512 : 'sha512');
+            case 'openssl':
+                if(defined(self::$supportedAlgorithms[$algo][2])){
+                    $algorithm = constant(self::$supportedAlgorithms[$algo][2]);
```

Comment:
can you explain exactly what's happening here? We're using the first array item as the algorithm, and the proceeding array items as arguments to the algorithm's function?